packages/core/source-map/src/SourceMap.js

```diff
@@ -0,0 +1,385 @@
+      name: mapping.name,
+      original: mapping.original,
+      generated: {
+        line: mapping.generated.line + lineOffset,
+        column: mapping.generated.column + columnOffset
+      }
+    });
+  }
+
+  addConsumerMapping(
+    mapping: ConsumerMappingItemType,
+    lineOffset: number = 0,
+    columnOffset: number = 0
+  ) {
+    let original = null;
+    if (
+      typeof mapping.originalLine === 'number' &&
+      mapping.originalLine > 0 &&
+      typeof mapping.originalColumn === 'number'
+    ) {
+      original = {
+        line: mapping.originalLine,
+        column: mapping.originalColumn
+      };
+    }
+
+    this.mappings.push({
+      source: original ? mapping.source : null,
+      name: mapping.name,
+      original,
+      generated: {
+        line: mapping.generatedLine + lineOffset,
+        column: mapping.generatedColumn + columnOffset
+      }
+    });
+  }
+
+  eachMapping(callback: (mapping: MappingType) => any) {
+    this.mappings.forEach(callback);
+  }
+
+  generateEmptyMap(sourceName: string, sourceContent: string) {
+    this.sources.set(sourceName, sourceContent);
+
+    let lineCount = lineCounter(sourceContent);
+    for (let line = 1; line < lineCount + 1; line++) {
+      this.addMapping({
+        source: sourceName,
+        original: {
+          line: line,
+          column: 0
+        },
+        generated: {
+          line: line,
+          column: 0
+        }
+      });
+    }
+
+    return this;
+  }
+
+  async extend(extension: SourceMap | RawMapInputType) {
+    if (!(extension instanceof SourceMap)) {
+      extension = await new SourceMap().addMap(extension);
+    }
+
+    return this._extend(extension);
+  }
+
+  async _extend(extension: SourceMap) {
+    extension.eachMapping(mapping => {
+      let originalMappingIndex = null;
+      if (mapping.original !== null) {
+        originalMappingIndex = this.findClosest(
+          mapping.original.line,
+          mapping.original.column
+        );
+      }
+
+      if (originalMappingIndex === null) {
+        this.addMapping(mapping);
+      } else {
+        let originalMapping = this.mappings[originalMappingIndex];
+        this.mappings[originalMappingIndex] = {
+          generated: mapping.generated,
+          original: originalMapping.original,
+          source: originalMapping.source,
+          name: originalMapping.name || mapping.name || null
+        };
+      }
+
+      if (mapping.source && !this.sources.has(mapping.source)) {
+        this.sources.set(
+          mapping.source,
+          extension.sourceContentFor(mapping.source)
+        );
+      }
+    });
+
+    return this;
+  }
+
+  findClosest(line: number, column: number): number | null {
+    if (line < 1) {
+      throw new Error('Line numbers must be >= 1');
+    }
+
+    if (column < 0) {
+      throw new Error('Column numbers must be >= 0');
+    }
+
+    if (this.mappings.length < 1) {
+      return null;
+    }
+
+    var startIndex = 0;
```

Comment:
`let`? 😄 (I know this is from Parcel 1, but we could update here while we're at it)

Reply:
Not sure if that matters, there were performance issues with let in v1 version

Reply:
Would like to benchmark these changes once I've got my hands on a bunch of sourcemaps I can run through this